src/lem/eval.rs

```diff
@@ -1,43 +1,148 @@
+    store::Store,
+    Ctrl, Func, Op, Tag, Var,
+};
 
 static EVAL_STEP: OnceCell<Func> = OnceCell::new();
 
-/// Lurk's step function
+/// Cached version of Lurk's default step function (IVC, no coprocessors)
+#[inline]
 pub fn eval_step() -> &'static Func {
-    EVAL_STEP.get_or_init(|| {
-        let reduce = reduce();
-        let apply_cont = apply_cont();
-        let make_thunk = make_thunk();
-
-        func!(step(expr, env, cont): 3 => {
-            let (expr, env, cont, ctrl) = reduce(expr, env, cont);
-            let (expr, env, cont, ctrl) = apply_cont(expr, env, cont, ctrl);
-            let (expr, env, cont, _ctrl) = make_thunk(expr, env, cont, ctrl);
-            return (expr, env, cont)
-        })
-    })
+    EVAL_STEP.get_or_init(|| make_eval_step(&[], true))
+}
+
+fn get_pc<F: LurkField, C: Coprocessor<F>>(
+    expr: &Ptr<F>,
+    store: &Store<F>,
+    lang: &Lang<F, C>,
+) -> usize {
+    match expr {
+        Ptr::Tuple2(Tag::Expr(Cproc), idx) => {
+            let (cproc, _) = store
+                .fetch_2_ptrs(*idx)
+                .expect("Coprocessor expression is not interned");
+            let cproc_sym = store
+                .fetch_symbol(cproc)
+                .expect("Coprocessor expression is not interned");
+            lang.get_index_by_symbol(&cproc_sym)
+                .expect("Coprocessor not found")
+                + 1
+        }
+        _ => 0,
+    }
+}
+
+// Builds frames for IVC or NIVC scheme
+pub fn build_frames<
+    F: LurkField,
+    C: Coprocessor<F>,
+    LogFmt: Fn(usize, &[Ptr<F>], &[Ptr<F>], &Store<F>) -> String,
+>(
+    lurk_step: &Func,
+    cprocs_run: &[Func],
+    mut input: Vec<Ptr<F>>,
+    store: &Store<F>,
+    limit: usize,
+    lang: &Lang<F, C>,
+    log_fmt: LogFmt,
+) -> Result<(Vec<Frame<F>>, usize)> {
+    let mut pc = 0;
+    let mut frames = vec![];
+    let mut iterations = 0;
+    tracing::info!("{}", &log_fmt(0, &input, &[], store));
+    for _ in 0..limit {
+        let func = if pc == 0 {
+            lurk_step
+        } else {
+            cprocs_run
+                .get(pc - 1)
+                .expect("Program counter outside range")
+        };
+        assert_eq!(func.input_params.len(), input.len());
+        let preimages = Preimages::new_from_func(func);
+        let mut emitted = vec![];
+        let (frame, _) = func.call(&input, store, preimages, &mut emitted, lang, pc)?;
+
+        iterations += 1;
+        input = frame.output.clone();
+        tracing::info!("{}", &log_fmt(iterations, &input, &emitted, store));
+        let must_break = matches!(frame.output[2].tag(), Tag::Cont(Terminal | Error));
+        let expr = frame.output[0];
+        frames.push(frame);
+
+        if must_break {
+            break;
+        }
+        pc = get_pc(&expr, store, lang);
+    }
+    Ok((frames, iterations))
+}
+
+/// Faster version of `build_frames` that doesn't accumulate frames
+pub fn traverse_frames<F: LurkField, C: Coprocessor<F>>(
```

Comment:
Where would this be used?

Reply:
This is used in most meta commands, for which proving doesn't matter and we can go faster.
Also, for testing evaluation.